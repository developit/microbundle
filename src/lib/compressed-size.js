import { basename } from 'path';
import { green, red, yellow, white } from 'kleur';
import gzipSize from 'gzip-size';
import brotliSize from 'brotli-size';
import prettyBytes from 'pretty-bytes';

function getPadLeft(str, width, char = ' ') {
	return char.repeat(width - str.length);
}

function formatSize(size, filename, type, raw) {
	const pretty = raw ? `${size} B` : prettyBytes(size);
	const color = size < 5000 ? green : size > 40000 ? red : yellow;
	const indent = getPadLeft(pretty, type === 'br' ? 13 : 10);
	return `${indent}${color(pretty)}: ${white(basename(filename))}.${type}`;
}

export async function getSizeInfo(code, filename, raw) {
	raw = raw || code.length < 5000;

	const [gzip, brotli] = await Promise.all([
		gzipSize(code).catch(() => null),
		brotliSize(code).catch(() => null),
	]);

	let out = formatSize(gzip, filename, 'gz', raw);
	if (brotli) {
		out += '\n' + formatSize(brotli, filename, 'br', raw);
	}

	return out;
}
