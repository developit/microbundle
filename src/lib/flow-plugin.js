import flow from 'rollup-plugin-flow';

export default function fixedFlow(options) {
	let plugin = flow(options);
	return Object.assign({}, plugin, {
		transform(code, id) {
			let ret = plugin.transform(code, id);
			if (ret.code===code) return null;
			return ret;
		}
	});
}