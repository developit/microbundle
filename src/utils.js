import { promises as fs } from 'fs';
import camelCase from 'camelcase';

export const readFile = fs.readFile;

export const stat = fs.stat;

export function isDir(name) {
	return stat(name)
		.then(stats => stats.isDirectory())
		.catch(() => false);
}

export function isFile(name) {
	return stat(name)
		.then(stats => stats.isFile())
		.catch(() => false);
}

// eslint-disable-next-line no-console
export const stdout = console.log.bind(console);
export const stderr = console.error.bind(console);

export const isTruthy = obj => {
	if (!obj) {
		return false;
	}

	return obj.constructor !== Object || Object.keys(obj).length > 0;
};

/** Remove a @scope/ prefix from a package name string */
export const removeScope = name => name.replace(/^@.*\//, '');

const INVALID_ES3_IDENT = /((^[^a-zA-Z]+)|[^\w.-])|([^a-zA-Z0-9]+$)/g;

/**
 * Turn a package name into a valid reasonably-unique variable name
 * @param {string} name
 */
export function safeVariableName(name) {
	const normalized = removeScope(name).toLowerCase();
	const identifier = normalized.replace(INVALID_ES3_IDENT, '');
	return camelCase(identifier);
}

export const EXTENSION = /(\.(umd|cjs|es|m))?\.([cm]?[tj]sx?)$/;
