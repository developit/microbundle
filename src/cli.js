#!/usr/bin/env node

import microbundle from '.';
import prog from './prog';
import { stdout } from './utils';
import logError from './log-error';

const run = opts => {
	microbundle(opts)
		.then(({ output }) => {
			if (output != null) stdout(output);
			if (!opts.watch) process.exit(0);
		})
		.catch(err => {
			process.exitCode = (typeof err.code === 'number' && err.code) || 1;
			logError(err);
			process.exit();
		});
};

prog(run)(process.argv);
