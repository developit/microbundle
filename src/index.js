import 'acorn-jsx';
import fs from 'fs';
import { resolve, relative, dirname, basename, extname } from 'path';
import chalk from 'chalk';
import { map, series } from 'asyncro';
import glob from 'tiny-glob/sync';
import autoprefixer from 'autoprefixer';
import cssnano from 'cssnano';
import { rollup, watch } from 'rollup';
import nodent from 'rollup-plugin-nodent';
import commonjs from 'rollup-plugin-commonjs';
import nodeResolve from 'rollup-plugin-node-resolve';
import buble from 'rollup-plugin-buble';
import { terser } from 'rollup-plugin-terser';
import postcss from 'rollup-plugin-postcss';
import alias from 'rollup-plugin-strict-alias';
import gzipSize from 'gzip-size';
import brotliSize from 'brotli-size';
import prettyBytes from 'pretty-bytes';
import shebangPlugin from 'rollup-plugin-preserve-shebang';
import typescript from 'rollup-plugin-typescript2';
import flow from './lib/flow-plugin';
import logError from './log-error';
import { readFile, isDir, isFile, stdout, stderr } from './utils';
import camelCase from 'camelcase';

const removeScope = name => name.replace(/^@.*\//, '');
const safeVariableName = name =>
	camelCase(
		removeScope(name)
			.toLowerCase()
			.replace(/((^[^a-zA-Z]+)|[^\w.-])|([^a-zA-Z0-9]+$)/g, ''),
	);
const parseGlobals = globalStrings => {
	const globals = {};
	globalStrings.split(',').forEach(globalString => {
		const [localName, globalName] = globalString.split('=');
		globals[localName] = globalName;
	});
	return globals;
};

const WATCH_OPTS = {
	exclude: 'node_modules/**',
};

export default async function microbundle(options) {
	let cwd = (options.cwd = resolve(process.cwd(), options.cwd)),
		hasPackageJson = true;

	try {
		options.pkg = JSON.parse(
			await readFile(resolve(cwd, 'package.json'), 'utf8'),
		);
	} catch (err) {
		stderr(
			chalk.yellow(
				`${chalk.yellow.inverse(
					'WARN',
				)} no package.json found. Assuming a pkg.name of "${basename(
					options.cwd,
				)}".`,
			),
		);
		let msg = String(err.message || err);
		if (!msg.match(/ENOENT/)) stderr(`  ${chalk.red.dim(msg)}`);
		options.pkg = {};
		hasPackageJson = false;
	}

	if (!options.pkg.name) {
		options.pkg.name = basename(options.cwd);
		if (hasPackageJson) {
			stderr(
				chalk.yellow(
					`${chalk.yellow.inverse(
						'WARN',
					)} missing package.json "name" field. Assuming "${
						options.pkg.name
					}".`,
				),
			);
		}
	}

	options.name =
		options.name || options.pkg.amdName || safeVariableName(options.pkg.name);

	const jsOrTs = async filename =>
		resolve(
			cwd,
			`${filename}${
				(await isFile(resolve(cwd, filename + '.ts')))
					? '.ts'
					: (await isFile(resolve(cwd, filename + '.tsx')))
						? '.tsx'
						: '.js'
			}`,
		);

	options.input = [];
	[]
		.concat(
			options.entries && options.entries.length
				? options.entries
				: (options.pkg.source && resolve(cwd, options.pkg.source)) ||
				  ((await isDir(resolve(cwd, 'src'))) && (await jsOrTs('src/index'))) ||
				  (await jsOrTs('index')) ||
				  options.pkg.module,
		)
		.map(file => glob(file))
		.forEach(file => options.input.push(...file));

	let main = resolve(cwd, options.output || options.pkg.main || 'dist');
	if (!main.match(/\.[a-z]+$/) || (await isDir(main))) {
		main = resolve(main, `${removeScope(options.pkg.name)}.js`);
	}
	options.output = main;

	let entries = (await map([].concat(options.input), async file => {
		file = resolve(cwd, file);
		if (await isDir(file)) {
			file = resolve(file, 'index.js');
		}
		return file;
	})).filter((item, i, arr) => arr.indexOf(item) === i);

	options.entries = entries;

	options.multipleEntries = entries.length > 1;

	let formats = (options.format || options.formats).split(',');
	// always compile cjs first if it's there:
	formats.sort((a, b) => (a === 'cjs' ? -1 : a > b ? 1 : 0));

	let steps = [];
	for (let i = 0; i < entries.length; i++) {
		for (let j = 0; j < formats.length; j++) {
			steps.push(
				createConfig(options, entries[i], formats[j], i === 0 && j === 0),
			);
		}
	}

	function formatSize(size, filename, type, raw) {
		const pretty = raw ? `${size} B` : prettyBytes(size);
		const color = size < 5000 ? 'green' : size > 40000 ? 'red' : 'yellow';
		const MAGIC_INDENTATION = type === 'br' ? 13 : 10;
		return `${' '.repeat(MAGIC_INDENTATION - pretty.length)}${chalk[color](
			pretty,
		)}: ${chalk.white(basename(filename))}.${type}`;
	}

	async function getSizeInfo(code, filename) {
		const raw = options.raw || code.length < 5000;
		const gzip = formatSize(await gzipSize(code), filename, 'gz', raw);
		const brotli = formatSize(await brotliSize(code), filename, 'br', raw);
		return gzip + '\n' + brotli;
	}

	if (options.watch) {
		const onBuild = options.onBuild;
		return new Promise((resolve, reject) => {
			stdout(
				chalk.blue(
					`Watching source, compiling to ${relative(
						cwd,
						dirname(options.output),
					)}:`,
				),
			);
			steps.map(options => {
				watch(
					Object.assign(
						{
							output: options.outputOptions,
							watch: WATCH_OPTS,
						},
						options.inputOptions,
					),
				).on('event', e => {
					if (e.code === 'FATAL') {
						return reject(e.error);
					} else if (e.code === 'ERROR') {
						logError(e.error);
					}
					if (e.code === 'END') {
						getSizeInfo(options._code, options.outputOptions.file).then(
							text => {
								stdout(`Wrote ${text.trim()}`);
							},
						);
						if (typeof onBuild === 'function') {
							onBuild(e);
						}
					}
				});
			});
		});
	}

	let cache;
	let out = await series(
		steps.map(({ inputOptions, outputOptions }) => async () => {
			inputOptions.cache = cache;
			let bundle = await rollup(inputOptions);
			cache = bundle;
			const { code } = await bundle.write(outputOptions);
			return await getSizeInfo(code, outputOptions.file);
		}),
	);

	return (
		chalk.blue(
			`Build "${options.name}" to ${relative(cwd, dirname(options.output)) ||
				'.'}:`,
		) +
		'\n   ' +
		out.join('\n   ')
	);
}

function createConfig(options, entry, format, writeMeta) {
	let { pkg } = options;

	let external = ['dns', 'fs', 'path', 'url'].concat(
		options.entries.filter(e => e !== entry),
	);

	let aliases = {};
	// since we transform src/index.js, we need to rename imports for it:
	if (options.multipleEntries) {
		aliases['.'] = './' + basename(options.output);
	}

	let useNodeResolve = true;
	const peerDeps = Object.keys(pkg.peerDependencies || {});
	if (options.external === 'none') {
		// bundle everything (external=[])
	} else if (options.external) {
		external = external.concat(peerDeps).concat(options.external.split(','));
	} else {
		external = external
			.concat(peerDeps)
			.concat(Object.keys(pkg.dependencies || {}));
	}

	let globals = external.reduce((globals, name) => {
		// valid JS identifiers are usually library globals:
		if (name.match(/^[a-z_$][a-z0-9_$]*$/)) {
			globals[name] = name;
		}
		return globals;
	}, {});
	if (options.globals && options.globals !== 'none') {
		globals = Object.assign(globals, parseGlobals(options.globals));
	}

	function replaceName(filename, name) {
		return resolve(
			dirname(filename),
			name + basename(filename).replace(/^[^.]+/, ''),
		);
	}

	let mainNoExtension = options.output;
	if (options.multipleEntries) {
		let name = entry.match(/([\\/])index(\.(umd|cjs|es|m))?\.m?js$/)
			? mainNoExtension
			: entry;
		mainNoExtension = resolve(dirname(mainNoExtension), basename(name));
	}
	mainNoExtension = mainNoExtension.replace(/(\.(umd|cjs|es|m))?\.m?js$/, '');

	let moduleMain = replaceName(
		pkg.module && !pkg.module.match(/src\//)
			? pkg.module
			: pkg['jsnext:main'] || 'x.mjs',
		mainNoExtension,
	);
	let cjsMain = replaceName(pkg['cjs:main'] || 'x.js', mainNoExtension);
	let umdMain = replaceName(pkg['umd:main'] || 'x.umd.js', mainNoExtension);

	// let rollupName = safeVariableName(basename(entry).replace(/\.js$/, ''));

	let nameCache = {};
	let mangleOptions = options.pkg.mangle || false;

	let exportType;
	if (format !== 'es') {
		try {
			let file = fs.readFileSync(entry, 'utf-8');
			let hasDefault = /\bexport\s*default\s*[a-zA-Z_$]/.test(file);
			let hasNamed =
				/\bexport\s*(let|const|var|async|function\*?)\s*[a-zA-Z_$*]/.test(
					file,
				) || /^\s*export\s*\{/m.test(file);
			if (hasDefault && hasNamed) exportType = 'default';
		} catch (e) {}
	}

	const useTypescript = extname(entry) === '.ts' || extname(entry) === '.tsx';

	const externalPredicate = new RegExp(`^(${external.join('|')})($|/)`);
	const externalTest =
		external.length === 0 ? () => false : id => externalPredicate.test(id);

	function loadNameCache() {
		try {
			nameCache = JSON.parse(
				fs.readFileSync(resolve(options.cwd, 'mangle.json'), 'utf8'),
			);
		} catch (e) {}
	}
	loadNameCache();

	let config = {
		inputOptions: {
			input: exportType ? resolve(__dirname, '../src/lib/__entry__.js') : entry,
			external: id => {
				if (options.multipleEntries && id === '.') {
					return true;
				}
				return externalTest(id);
			},
			plugins: []
				.concat(
					alias({
						__microbundle_entry__: entry,
					}),
					postcss({
						plugins: [
							autoprefixer(),
							options.compress !== false &&
								cssnano({
									preset: 'default',
								}),
						].filter(Boolean),
						// only write out CSS for the first bundle (avoids pointless extra files):
						inject: false,
						extract: !!writeMeta,
					}),
					useTypescript &&
						typescript({
							typescript: require('typescript'),
							cacheRoot: `./.rts2_cache_${format}`,
							tsconfigDefaults: {
								compilerOptions: { declaration: true, jsx: options.jsx },
							},
						}),
					!useTypescript && flow({ all: true, pretty: true }),
					nodent({
						exclude: 'node_modules/**',
						noRuntime: true,
						promises: true,
						transformations: {
							forOf: false,
						},
						parser: {
							plugins: {
								jsx: true,
							},
						},
					}),
					!useTypescript &&
						buble({
							exclude: 'node_modules/**',
							jsx: options.jsx || 'h',
							objectAssign: options.assign || 'Object.assign',
							transforms: {
								dangerousForOf: true,
								dangerousTaggedTemplateString: true,
							},
						}),
					useNodeResolve &&
						commonjs({
							include: 'node_modules/**',
						}),
					useNodeResolve &&
						nodeResolve({
							module: true,
							jsnext: true,
							browser: options.target !== 'node',
						}),
					// We should upstream this to rollup
					// format==='cjs' && replace({
					// 	[`module.exports = ${rollupName};`]: '',
					// 	[`var ${rollupName} =`]: 'module.exports ='
					// }),
					// This works for the general case, but could cause nasty scope bugs.
					// format==='umd' && replace({
					// 	[`return ${rollupName};`]: '',
					// 	[`var ${rollupName} =`]: 'return'
					// }),
					// format==='es' && replace({
					// 	[`export default ${rollupName};`]: '',
					// 	[`var ${rollupName} =`]: 'export default'
					// }),
					options.compress !== false && [
						terser({
							sourceMap: true,
							output: { comments: false },
							compress: {
								keep_infinity: true,
								pure_getters: true,
							},
							warnings: true,
							ecma: 5,
							toplevel: format === 'cjs' || format === 'es',
							mangle: {
								properties: mangleOptions
									? {
											regex: mangleOptions.regex
												? new RegExp(mangleOptions.regex)
												: null,
											reserved: mangleOptions.reserved || [],
									  }
									: false,
							},
							nameCache,
						}),
						mangleOptions && {
							// before hook
							options: loadNameCache,
							// after hook
							onwrite() {
								if (writeMeta && nameCache) {
									fs.writeFile(
										resolve(options.cwd, 'mangle.json'),
										JSON.stringify(nameCache, null, 2),
										Object,
									);
								}
							},
						},
					],
					{
						ongenerate(outputOptions, { code }) {
							config._code = code;
						},
					},
					shebangPlugin(),
				)
				.filter(Boolean),
		},

		outputOptions: {
			exports: exportType ? 'default' : undefined,
			paths: aliases,
			globals,
			strict: options.strict === true,
			legacy: true,
			freeze: false,
			esModule: false,
			sourcemap: options.sourcemap !== false,
			treeshake: {
				propertyReadSideEffects: false,
			},
			format,
			name: options.name,
			file: resolve(
				options.cwd,
				(format === 'es' && moduleMain) ||
					(format === 'umd' && umdMain) ||
					cjsMain,
			),
		},
	};

	return config;
}
