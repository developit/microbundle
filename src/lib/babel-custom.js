import { createConfigItem } from '@babel/core';
import babelPlugin from 'rollup-plugin-babel';
import merge from 'lodash.merge';
import { isTruthy } from '../utils';

const ESMODULES_TARGET = {
	esmodules: true,
};

const mergeConfigItems = (type, ...configItemsToMerge) => {
	const mergedItems = [];

	configItemsToMerge.forEach(configItemToMerge => {
		configItemToMerge.forEach(item => {
			const itemToMergeWithIndex = mergedItems.findIndex(
				mergedItem => mergedItem.file.resolved === item.file.resolved,
			);

			if (itemToMergeWithIndex === -1) {
				mergedItems.push(item);
				return;
			}

			mergedItems[itemToMergeWithIndex] = createConfigItem(
				[
					mergedItems[itemToMergeWithIndex].file.resolved,
					merge(mergedItems[itemToMergeWithIndex].options, item.options),
				],
				{
					type,
				},
			);
		});
	});

	return mergedItems;
};

const createConfigItems = (type, items) => {
	return items.map(({ name, ...options }) => {
		return createConfigItem([require.resolve(name), options], { type });
	});
};

const presetEnvRegex = RegExp(/@babel\/(preset-)?env/);

export default () => {
	return babelPlugin.custom(babelCore => {
		return {
			// Passed the plugin options.
			options({ custom: customOptions, ...pluginOptions }) {
				return {
					// Pull out any custom options that the plugin might have.
					customOptions,

					// Pass the options back with the two custom options removed.
					pluginOptions,
				};
			},

			config(config, { customOptions }) {
				const defaultPlugins = createConfigItems(
					'plugin',
					[
						{
							name: '@babel/plugin-transform-react-jsx',
							pragma: customOptions.pragma || 'h',
							pragmaFrag: customOptions.pragmaFrag || 'Fragment',
						},
						!customOptions.typescript && {
							name: '@babel/plugin-transform-flow-strip-types',
						},
						isTruthy(customOptions.defines) && {
							name: 'babel-plugin-transform-replace-expressions',
							replace: customOptions.defines,
						},
						!customOptions.modern && {
							name: 'babel-plugin-transform-async-to-promises',
							inlineHelpers: true,
							externalHelpers: true,
						},
						{
							name: '@babel/plugin-proposal-class-properties',
							loose: true,
						},
						!customOptions.modern && {
							name: '@babel/plugin-transform-regenerator',
							async: false,
						},
						{
							name: 'babel-plugin-macros',
						},
					].filter(Boolean),
				);

				const babelOptions = config.options || {};

				const envIdx = (babelOptions.presets || []).findIndex(preset =>
					presetEnvRegex.test(preset.file.request),
				);

				const environmentPreset = customOptions.modern
					? '@babel/preset-modules'
					: '@babel/preset-env';

				if (envIdx !== -1) {
					const preset = babelOptions.presets[envIdx];
					babelOptions.presets[envIdx] = createConfigItem(
						[
							environmentPreset,
							Object.assign(
								merge(
									{
										loose: true,
										useBuiltIns: false,
										targets: customOptions.targets,
									},
									preset.options,
									{
										modules: false,
										exclude: merge(
											['transform-async-to-generator', 'transform-regenerator'],
											(preset.options && preset.options.exclude) || [],
										),
									},
								),
								customOptions.modern ? { targets: ESMODULES_TARGET } : {},
							),
						],
						{
							type: `preset`,
						},
					);
				} else {
					babelOptions.presets = createConfigItems('preset', [
						{
							name: environmentPreset,
							targets: customOptions.modern
								? ESMODULES_TARGET
								: customOptions.targets,
							modules: false,
							loose: true,
							useBuiltIns: false,
							exclude: [
								'transform-async-to-generator',
								'transform-regenerator',
							],
						},
					]);
				}

				// Merge babelrc & our plugins together
				babelOptions.plugins = mergeConfigItems(
					'plugin',
					defaultPlugins,
					babelOptions.plugins || [],
				);

				babelOptions.generatorOpts = {
					minified: customOptions.compress,
					compact: customOptions.compress,
					shouldPrintComment: comment => /[@#]__PURE__/.test(comment),
				};

				return babelOptions;
			},
		};
	});
};
