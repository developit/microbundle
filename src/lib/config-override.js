import { isFile } from '../utils';
import { resolve } from 'path';

function configOverrider(projectOverride, context) {
	return {
		/**
		 * Override the configuration for a given plugin.
		 *
		 * @param {string} pluginName
		 * @param {Object} config the
		 */
		pluginConfig(pluginName, config) {
			// No override if no plugins override is defined
			if (!projectOverride.plugins) {
				return config;
			}

			// Expect provided override to be a function returning modified config
			const override = projectOverride.plugins[pluginName];
			return override ? override(config, context) : config;
		},

		/**
		 * Override the full rollup config before it's used
		 *
		 * @param {Object} config
		 */
		config(config) {
			if (!projectOverride.config) {
				return config;
			}

			return projectOverride.config(config, context);
		},
	};
}

export async function getConfigOverride(context) {
	const path = resolve(context.options.cwd, 'microbundle.config.js');
	const hasProjectConfig = await isFile(path);

	return configOverrider(hasProjectConfig ? require(path) : {}, context);
}
