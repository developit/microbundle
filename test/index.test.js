import { resolve } from 'path';
import fs from 'fs-extra';
import { promisify } from 'es6-promisify';
import dirTree from 'directory-tree';
import shellQuote from 'shell-quote';
import _rimraf from 'rimraf';
import { strip } from './lib/util';
import { readFile } from '../src/utils';
import createProg from '../src/prog';
import microbundle from '../src/index';

const rimraf = promisify(_rimraf);

const FIXTURES_DIR = `${__dirname}/fixtures`;
const DEFAULT_SCRIPT = 'microbundle';

const join = (arr, delimiter = '') => arr.join(delimiter);

const printTree = (nodes, indentLevel = 0) => {
	const indent = '  '.repeat(indentLevel);
	return join(
		nodes
			.filter(node => node.name[0] !== '.')
			.map(node => {
				const isDir = node.type === 'directory';
				return `${indent}${node.name}\n${
					isDir ? printTree(node.children, indentLevel + 1) : ''
				}`;
			}),
	);
};

const getBuildScript = async (fixturePath, defaultScript) => {
	let pkg = {};
	try {
		pkg = JSON.parse(
			await readFile(resolve(fixturePath, 'package.json'), 'utf8'),
		);
	} catch (err) {
		if (err.code !== 'ENOENT') throw err;
	}
	return (pkg && pkg.scripts && pkg.scripts.build) || defaultScript;
};

const parseScript = (() => {
	let parsed;
	const prog = createProg(_parsed => (parsed = _parsed));
	return script => {
		const argv = shellQuote.parse(`node ${script}`);
		// assuming {op: 'glob', pattern} for non-string args
		prog(argv.map(arg => (typeof arg === 'string' ? arg : arg.pattern)));
		return parsed;
	};
})();

describe('fixtures', () => {
	fs.readdirSync(FIXTURES_DIR).forEach(fixtureDir => {
		let fixturePath = resolve(FIXTURES_DIR, fixtureDir);

		if (!fs.statSync(fixturePath).isDirectory()) {
			return;
		}

		it(fixtureDir, async () => {
			if (fixtureDir.endsWith('-with-cwd')) {
				fixturePath = resolve(fixturePath, fixtureDir.replace('-with-cwd', ''));
			}

			const dist = resolve(`${fixturePath}/dist`);
			// clean up
			await rimraf(dist);
			await rimraf(resolve(`${fixturePath}/.rts2_cache_cjs`));
			await rimraf(resolve(`${fixturePath}/.rts2_cache_es`));
			await rimraf(resolve(`${fixturePath}/.rts2_cache_umd`));

			const script = await getBuildScript(fixturePath, DEFAULT_SCRIPT);

			const prevDir = process.cwd();
			process.chdir(resolve(fixturePath));

			const parsedOpts = parseScript(script);

			const output = await microbundle({
				...parsedOpts,
				cwd: parsedOpts.cwd !== '.' ? parsedOpts.cwd : resolve(fixturePath),
			});

			process.chdir(prevDir);

			const printedDir = printTree([dirTree(fixturePath)]);

			expect(
				[
					`Used script: ${script}`,
					'Directory tree:',
					printedDir,
					strip(output),
				].join('\n\n'),
			).toMatchSnapshot();

			fs.readdirSync(resolve(dist)).forEach(file => {
				expect(
					fs.readFileSync(resolve(dist, file)).toString('utf8'),
				).toMatchSnapshot();
			});
		});
	});

	it('should keep shebang', () => {
		expect(
			fs
				.readFileSync(resolve(FIXTURES_DIR, 'shebang/dist/shebang.js'), 'utf8')
				.startsWith('#!'),
		).toEqual(true);
	});

	it('should keep named and default export', () => {
		const mod = require(resolve(
			FIXTURES_DIR,
			'default-named/dist/default-named.js',
		));

		expect(Object.keys(mod)).toEqual(['foo', 'default']);
	});
});
