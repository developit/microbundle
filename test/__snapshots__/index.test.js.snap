// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixtures basic 1`] = `
"Used script: microbundle

Directory tree:

basic
  dist
    basic-lib.js
    basic-lib.js.map
    basic-lib.mjs
    basic-lib.mjs.map
    basic-lib.umd.js
    basic-lib.umd.js.map
  package.json
  src
    index.js
    two.js


Build \\"basicLib\\" to dist:
211 B: basic-lib.js.gz
154 B: basic-lib.js.br
210 B: basic-lib.mjs.gz
155 B: basic-lib.mjs.br
292 B: basic-lib.umd.js.gz
228 B: basic-lib.umd.js.br"
`;

exports[`fixtures basic-css 1`] = `
"Used script: microbundle

Directory tree:

basic-css
  dist
    basic-css.css
    basic-css.js
    basic-css.js.map
    basic-css.mjs
    basic-css.mjs.map
    basic-css.umd.js
    basic-css.umd.js.map
  package.json
  src
    index.js
    two.css


Build \\"basicCss\\" to dist:
107 B: basic-css.js.gz
60 B: basic-css.js.br
106 B: basic-css.mjs.gz
56 B: basic-css.mjs.br
188 B: basic-css.umd.js.gz
135 B: basic-css.umd.js.br"
`;

exports[`fixtures basic-ts 1`] = `
"Used script: microbundle

Directory tree:

basic-ts
  dist
    basic-lib-ts.js
    basic-lib-ts.js.map
    basic-lib-ts.mjs
    basic-lib-ts.mjs.map
    basic-lib-ts.umd.js
    basic-lib-ts.umd.js.map
    car.d.ts
    index.d.ts
  package.json
  src
    car.ts
    index.ts
  tsconfig.json


Build \\"basicLibTs\\" to dist:
104 B: basic-lib-ts.js.gz
76 B: basic-lib-ts.js.br
97 B: basic-lib-ts.mjs.gz
73 B: basic-lib-ts.mjs.br
180 B: basic-lib-ts.umd.js.gz
137 B: basic-lib-ts.umd.js.br"
`;

exports[`fixtures basic-tsx 1`] = `
"Used script: microbundle

Directory tree:

basic-tsx
  dist
    basic-lib-tsx.js
    basic-lib-tsx.js.map
    basic-lib-tsx.mjs
    basic-lib-tsx.mjs.map
    basic-lib-tsx.umd.js
    basic-lib-tsx.umd.js.map
    index.d.ts
  package.json
  src
    index.tsx
  tsconfig.json


Build \\"basicLibTsx\\" to dist:
221 B: basic-lib-tsx.js.gz
179 B: basic-lib-tsx.js.br
222 B: basic-lib-tsx.mjs.gz
197 B: basic-lib-tsx.mjs.br
295 B: basic-lib-tsx.umd.js.gz
236 B: basic-lib-tsx.umd.js.br"
`;

exports[`fixtures basic-with-cwd 1`] = `
"Used script: microbundle --cwd ./basic

Directory tree:

basic-with-cwd
  basic
    dist
      basic.js
      basic.js.map
      basic.mjs
      basic.mjs.map
      basic.umd.js
      basic.umd.js.map
    package.json
    src
      index.js
      two.js
  package.json


Build \\"basic\\" to dist:
211 B: basic.js.gz
154 B: basic.js.br
210 B: basic.mjs.gz
155 B: basic.mjs.br
287 B: basic.umd.js.gz
224 B: basic.umd.js.br"
`;

exports[`fixtures custom-source 1`] = `
"Used script: microbundle

Directory tree:

custom-source
  dist
    custom-source.js
    custom-source.js.map
    custom-source.mjs
    custom-source.mjs.map
    custom-source.umd.js
    custom-source.umd.js.map
  package.json
  src
    custom-source.js
    two.js


Build \\"customSource\\" to dist:
211 B: custom-source.js.gz
154 B: custom-source.js.br
210 B: custom-source.mjs.gz
155 B: custom-source.mjs.br
293 B: custom-source.umd.js.gz
226 B: custom-source.umd.js.br"
`;

exports[`fixtures custom-source-with-cwd 1`] = `
"Used script: microbundle --cwd ./custom-source

Directory tree:

custom-source-with-cwd
  custom-source
    dist
      custom-src.js
      custom-src.js.map
      custom-src.mjs
      custom-src.mjs.map
      custom-src.umd.js
      custom-src.umd.js.map
    package.json
    src
      custom-source.js
      two.js
  package.json


Build \\"customSrc\\" to dist:
211 B: custom-src.js.gz
154 B: custom-src.js.br
210 B: custom-src.mjs.gz
155 B: custom-src.mjs.br
291 B: custom-src.umd.js.gz
228 B: custom-src.umd.js.br"
`;

exports[`fixtures jsx 1`] = `
"Used script: microbundle

Directory tree:

jsx
  dist
    jsx.js
    jsx.js.map
    jsx.mjs
    jsx.mjs.map
    jsx.umd.js
    jsx.umd.js.map
  index.js
  package.json


Build \\"jsx\\" to dist:
211 B: jsx.js.gz
168 B: jsx.js.br
212 B: jsx.mjs.gz
170 B: jsx.mjs.br
283 B: jsx.umd.js.gz
230 B: jsx.umd.js.br"
`;

exports[`fixtures name-custom-amd 1`] = `
"Used script: microbundle

Directory tree:

name-custom-amd
  dist
    name-custom-amd.js
    name-custom-amd.js.map
    name-custom-amd.mjs
    name-custom-amd.mjs.map
    name-custom-amd.umd.js
    name-custom-amd.umd.js.map
  package.json
  src
    index.js
    two.js


Build \\"customNameAmd\\" to dist:
211 B: name-custom-amd.js.gz
154 B: name-custom-amd.js.br
210 B: name-custom-amd.mjs.gz
155 B: name-custom-amd.mjs.br
295 B: name-custom-amd.umd.js.gz
230 B: name-custom-amd.umd.js.br"
`;

exports[`fixtures name-custom-cli 1`] = `
"Used script: microbundle --name nameCustomCli

Directory tree:

name-custom-cli
  dist
    name-custom.js
    name-custom.js.map
    name-custom.mjs
    name-custom.mjs.map
    name-custom.umd.js
    name-custom.umd.js.map
  package.json
  src
    index.js
    two.js


Build \\"nameCustomCli\\" to dist:
211 B: name-custom.js.gz
154 B: name-custom.js.br
210 B: name-custom.mjs.gz
155 B: name-custom.mjs.br
294 B: name-custom.umd.js.gz
231 B: name-custom.umd.js.br"
`;

exports[`fixtures no-pkg 1`] = `
"Used script: microbundle

Directory tree:

no-pkg
  dist
    no-pkg.js
    no-pkg.js.map
    no-pkg.mjs
    no-pkg.mjs.map
    no-pkg.umd.js
    no-pkg.umd.js.map
  src
    index.js
    two.js


Build \\"noPkg\\" to dist:
211 B: no-pkg.js.gz
154 B: no-pkg.js.br
210 B: no-pkg.mjs.gz
155 B: no-pkg.mjs.br
288 B: no-pkg.umd.js.gz
228 B: no-pkg.umd.js.br"
`;

exports[`fixtures no-pkg-name 1`] = `
"Used script: microbundle

Directory tree:

no-pkg-name
  dist
    no-pkg-name.js
    no-pkg-name.js.map
    no-pkg-name.mjs
    no-pkg-name.mjs.map
    no-pkg-name.umd.js
    no-pkg-name.umd.js.map
  package.json
  src
    index.js
    two.js


Build \\"noPkgName\\" to dist:
211 B: no-pkg-name.js.gz
154 B: no-pkg-name.js.br
210 B: no-pkg-name.mjs.gz
155 B: no-pkg-name.mjs.br
292 B: no-pkg-name.umd.js.gz
230 B: no-pkg-name.umd.js.br"
`;

exports[`fixtures pretty 1`] = `
"Used script: microbundle

Directory tree:

pretty
  dist
    pretty.js
    pretty.js.map
    pretty.mjs
    pretty.mjs.map
    pretty.umd.js
    pretty.umd.js.map
  package.json
  src
    index.js


Build \\"pretty\\" to dist:
6.33 kB: pretty.js.gz
5.19 kB: pretty.js.br
6.34 kB: pretty.mjs.gz
5.21 kB: pretty.mjs.br
6.46 kB: pretty.umd.js.gz
5.3 kB: pretty.umd.js.br"
`;

exports[`fixtures raw 1`] = `
"Used script: microbundle --raw

Directory tree:

raw
  dist
    raw.js
    raw.js.map
    raw.mjs
    raw.mjs.map
    raw.umd.js
    raw.umd.js.map
  package.json
  src
    index.js


Build \\"raw\\" to dist:
1389 B: raw.js.gz
1160 B: raw.js.br
1398 B: raw.mjs.gz
1169 B: raw.mjs.br
1493 B: raw.umd.js.gz
1245 B: raw.umd.js.br"
`;
